### PR TITLE
Simplify our global new/delete implementations

### LINKED ISSUE #101708: superpmi tool crashing with "Heap contamination detected!"
This assert happened in [runtime-coreclr superpmi-collect](https://dev.azure.com/dnceng/internal/_build?definitionId=977&_a=summary) pipeline

```
[18:03:51] Invoking: /tmp/helix/working/AF7F0923/p/coreclr/superpmi -p -f /tmp/helix/working/AF7F0923/w/B2C509AA/u/spmi_collect/basefail.mcl /tmp/helix/working/AF7F0923/w/B2C509AA/u/spmi_collect/base.mch /tmp/helix/working/AF7F0923/p/coreclr/libclrjit.dylib
[18:04:10] 
[18:04:10] Assert failure(PID 25678 [0x0000644e], Thread: 278306 [0x43f22]): !"Heap contamination detected! HeapFree was called on a heap other than the one that memory was allocated from.\n" "Possible cause: you used new (executable) to allocate the memory, but didn't use DeleteExecutable() to free it."
[18:04:10]     File: /Users/runner/work/1/s/src/coreclr/utilcode/clrhost_nodependencies.cpp:290
[18:04:10]     Image: /private/tmp/helix/working/AF7F0923/p/coreclr/superpmi
```

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Simplify our global overrides of new/delete and remove quite a bit of dead code.

- Stack Hashing to track allocations was a no-op as it depended on a debug-only .NET Framework component, so even when the code was enabled, it did nothing. Remove this code.
- Our debug tagging doesn't get us much now that we have AddressSanitizer runs that can accurately detect exactly when heap corruption happens. Remove it. Fixes the symptom of https://github.com/dotnet/runtime/issues/101708.
- The `new(executable)` variants of `new` are not used anywhere in the code-base and they won't be now that we support W/X.
- Since VS2012, the Windows CRT has used the exact same allocation implementation that we were using (the process heap). Just use that and don't maintain the code ourselves.

I'm confident that this fixes https://github.com/dotnet/runtime/issues/101708 completely as all of the C++ standard libraries we use call malloc/free under the hood the same way we do (and macOS's libc++ in particular does).

There are two follow-up changes we can make that will allow us to fully remove our global new/delete overrides:

- Set up the StressLog OOM logging in CoreCLR using [`set_new_handler`](https://en.cppreference.com/w/cpp/memory/new/set_new_handler).
- Update CoreCLR's exception handling macros to handle [`std::bad_alloc`](https://en.cppreference.com/w/cpp/memory/new/bad_alloc) and translate it to our pre-allocated `OutOfMemoryException` object.

I plan to explore these in future PRs.

### CODE REVIEW

File: src/coreclr/utilcode/clrhost_nodependencies.cpp [hunk 1]
```diff
@@ -307,7 +242,7 @@ operator new(size_t n)
     STATIC_CONTRACT_FAULT;
     STATIC_CONTRACT_SUPPORTS_DAC_HOST_ONLY;
 
-    void* result = ClrMalloc(n);
+    void* result = malloc(n == 0 ? 1 : n);
     if (result == NULL) {
         ThrowOutOfMemory();
```
Comment (jkotas) [inline src/coreclr/utilcode/clrhost_nodependencies.cpp:247]: ClrMalloc did the OOM stress logging here before. Add it back?

File: src/coreclr/utilcode/clrhost_nodependencies.cpp [hunk 2]
```diff
@@ -204,96 +204,31 @@ ClrDebugState *CLRInitDebugState()
...
-    HANDLE hHeap = g_hProcessHeap;
-    if (hHeap == NULL)
+    void ReportOOM(size_t size)
     {
-        InterlockedCompareExchangeT(&g_hProcessHeap, ::GetProcessHeap(), NULL);
-        hHeap = g_hProcessHeap;
+        // With no host, we have nowhere to report the OOM.
     }
-
-    p = HeapAlloc(hHeap, 0, size);
```
Comment (jkotas) [inline src/coreclr/utilcode/clrhost_nodependencies.cpp:243]: How much more expensive is malloc compared to HeapAlloc on Windows?

File: src/coreclr/utilcode/clrhost_nodependencies.cpp [hunk 3]
```diff
@@ -318,39 +253,10 @@ operator new(size_t n)
...
-    STATIC_CONTRACT_GC_NOTRIGGER;
-    STATIC_CONTRACT_FAULT;
-    STATIC_CONTRACT_SUPPORTS_DAC_HOST_ONLY;
-
-    INCONTRACT(_ASSERTE(!ARE_FAULTS_FORBIDDEN()));
-
-    void* result = ClrMalloc(n);
-	TRASH_LASTERROR;
-    return result;
+    return ::operator new(n);
```
Comment (jkotas) [inline src/coreclr/utilcode/clrhost_nodependencies.cpp:256]: I am wondering why this was manually inlined before. Is this wrapper going to introduce an extra path length?

**Review by jkotas — COMMENTED** [review]
A few questions, LGTM otherwise. Thanks!

File: src/coreclr/utilcode/clrhost_nodependencies.cpp (hunk 2 above)
Comment (jkoritzinsky) [reply src/coreclr/utilcode/clrhost_nodependencies.cpp:243]: Nowadays, `malloc` just calls `HeapAlloc`, so it should be minimal extra cost. Maybe a few instructions for a tail call or a well-predicted if check (checking if we're using the system heap, which is the default).

File: src/coreclr/utilcode/clrhost_nodependencies.cpp (hunk 3 above)
Comment (jkoritzinsky) [reply src/coreclr/utilcode/clrhost_nodependencies.cpp:256]: I would expect this to be inlined or tail-called with our optimization settings.

File: src/coreclr/utilcode/clrhost_nodependencies.cpp (hunk 2 above)
Comment (jkotas) [reply src/coreclr/utilcode/clrhost_nodependencies.cpp:243]: I have done a quick perf check on Winx64 and I do not see a measurable impact from this change.

**Review by jkotas — APPROVED** [review]
Thanks

**Pull request merged**